# Introduction

This project analyzes user behavior and conversion patterns in an e-commerce product, focusing on the user journey from add_to_cart to purchase.

The goal is to identify key drop-off points in the funnel and understand which factors influence conversion at different stages.

## Data Overview

Synthetic E-commerce Transactions + Clickstream (2020–2025)
The dataset represents a synthetic e-commerce environment with multiple interconnected tables, simulating real-world user behavior and purchase patterns.

The dataset includes the following tables:

- customers: user information
- sessions: user sessions (device, source, time)
- events: user interactions within sessions
- orders: completed purchases
  
#### Tables

customers.csv — customer profiles, signup dates, country, opt-in

products.csv — catalog with categories, prices, costs, margins

sessions.csv — session metadata (device, source, start time, country)

events.csv — page_view / add_to_cart / checkout / purchase with timestamps

orders.csv — order headers (payment, discount, totals)

order_items.csv — line items (quantity, unit price, line total)

reviews.csv — product ratings & short text reviews



In [3]:
def create_connection():
    try:
        engine = create_engine(
            "mysql+pymysql://test_user:1234@127.0.0.1:3306/ecommerce_app"
        )
        
        with engine.connect() as conn:
            conn.execute(text("SELECT 1"))

        print("✅ Підключення до БД успішне!")
        return engine

    except Exception as e:
        print(f"❌ Помилка підключення: {e}")
        return None

engine = create_connection()

✅ Підключення до БД успішне!


### Data Quality Check

In [4]:
query_data_quality = """
-- CUSTOMERS
-- CUSTOMER_ID
SELECT
    'customers' AS table_name,
    'customer_id' AS column_name,
    COUNT(*) AS row_count,
    SUM(customer_id IS NULL) AS null_count,
    0 AS empty_count,
    ROUND(100.0 * SUM(customer_id IS NULL) / COUNT(*), 2) AS missing_pct,
    (
        SELECT COUNT(*)
        FROM (
            SELECT customer_id
            FROM customers
            GROUP BY customer_id
            HAVING COUNT(*) > 1
        ) t
    ) AS duplicate_count
FROM customers

UNION ALL

-- SIGNUP_DATE
SELECT
    'customers',
    'signup_date',
    COUNT(*),
    SUM(signup_date IS NULL),
    0,
    ROUND(100.0 * SUM(signup_date IS NULL) / COUNT(*), 2),
    NULL
FROM customers

UNION ALL

-- COUNTRY
SELECT
    'customers',
    'country',
    COUNT(*),
    SUM(country IS NULL),
    SUM(TRIM(COALESCE(country, '')) = ''),
    ROUND(
        100.0 * (
            SUM(country IS NULL) +
            SUM(TRIM(COALESCE(country, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM customers

UNION ALL

-- SESSIONS
-- SESSION_ID
SELECT
    'sessions',
    'session_id',
    COUNT(*),
    SUM(session_id IS NULL),
    0,
    ROUND(100.0 * SUM(session_id IS NULL) / COUNT(*), 2),
    (
        SELECT COUNT(*)
        FROM (
            SELECT session_id
            FROM sessions
            GROUP BY session_id
            HAVING COUNT(*) > 1
        ) t
    )
FROM sessions

UNION ALL

-- CUSTOMER_ID
SELECT
    'sessions',
    'customer_id',
    COUNT(*),
    SUM(customer_id IS NULL),
    0,
    ROUND(100.0 * SUM(customer_id IS NULL) / COUNT(*), 2),
    NULL
FROM sessions

UNION ALL

-- START_TIME
SELECT
    'sessions',
    'start_time',
    COUNT(*),
    SUM(start_time IS NULL),
    0,
    ROUND(100.0 * SUM(start_time IS NULL) / COUNT(*), 2),
    NULL
FROM sessions

UNION ALL

-- DEVICE
SELECT
    'sessions',
    'device',
    COUNT(*),
    SUM(device IS NULL),
    SUM(TRIM(COALESCE(device, '')) = ''),
    ROUND(
        100.0 * (
            SUM(device IS NULL) +
            SUM(TRIM(COALESCE(device, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM sessions

UNION ALL

-- SOURCE
SELECT
    'sessions',
    'source',
    COUNT(*),
    SUM(source IS NULL),
    SUM(TRIM(COALESCE(source, '')) = ''),
    ROUND(
        100.0 * (
            SUM(source IS NULL) +
            SUM(TRIM(COALESCE(source, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM sessions

UNION ALL

-- COUNTRY
SELECT
    'sessions',
    'country',
    COUNT(*),
    SUM(country IS NULL),
    SUM(TRIM(COALESCE(country, '')) = ''),
    ROUND(
        100.0 * (
            SUM(country IS NULL) +
            SUM(TRIM(COALESCE(country, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM sessions

UNION ALL

-- EVENTS
-- EVENT_ID
SELECT
    'events',
    'event_id',
    COUNT(*),
    SUM(event_id IS NULL),
    0,
    ROUND(100.0 * SUM(event_id IS NULL) / COUNT(*), 2),
    (
        SELECT COUNT(*)
        FROM (
            SELECT event_id
            FROM events
            GROUP BY event_id
            HAVING COUNT(*) > 1
        ) t
    )
FROM events

UNION ALL

-- SESSION_ID
SELECT
    'events',
    'session_id',
    COUNT(*),
    SUM(session_id IS NULL),
    0,
    ROUND(100.0 * SUM(session_id IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- EVENT_TIME
SELECT
    'events',
    'event_time',
    COUNT(*),
    SUM(event_time IS NULL),
    0,
    ROUND(100.0 * SUM(event_time IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- EVENT_TYPE
SELECT
    'events',
    'event_type',
    COUNT(*),
    SUM(event_type IS NULL),
    SUM(TRIM(COALESCE(event_type, '')) = ''),
    ROUND(
        100.0 * (
            SUM(event_type IS NULL) +
            SUM(TRIM(COALESCE(event_type, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM events

UNION ALL

-- PRODUCT_ID
SELECT
    'events',
    'product_id',
    COUNT(*),
    SUM(product_id IS NULL),
    0,
    ROUND(100.0 * SUM(product_id IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- QTY
SELECT
    'events',
    'qty',
    COUNT(*),
    SUM(qty IS NULL),
    0,
    ROUND(100.0 * SUM(qty IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- CART_SIZE
SELECT
    'events',
    'cart_size',
    COUNT(*),
    SUM(cart_size IS NULL),
    0,
    ROUND(100.0 * SUM(cart_size IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- PAYMENT
SELECT
    'events',
    'payment',
    COUNT(*),
    SUM(payment IS NULL),
    SUM(TRIM(COALESCE(payment, '')) = ''),
    ROUND(
        100.0 * (
            SUM(payment IS NULL) +
            SUM(TRIM(COALESCE(payment, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM events

UNION ALL

-- DISCOUNT_PCT
SELECT
    'events',
    'discount_pct',
    COUNT(*),
    SUM(discount_pct IS NULL),
    0,
    ROUND(100.0 * SUM(discount_pct IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- AMOUNT_USD
SELECT
    'events',
    'amount_usd',
    COUNT(*),
    SUM(amount_usd IS NULL),
    0,
    ROUND(100.0 * SUM(amount_usd IS NULL) / COUNT(*), 2),
    NULL
FROM events

UNION ALL

-- ORDERS
-- ORDER_ID
SELECT
    'orders',
    'order_id',
    COUNT(*),
    SUM(order_id IS NULL),
    0,
    ROUND(100.0 * SUM(order_id IS NULL) / COUNT(*), 2),
    (
        SELECT COUNT(*)
        FROM (
            SELECT order_id
            FROM orders
            GROUP BY order_id
            HAVING COUNT(*) > 1
        ) t
    )
FROM orders

UNION ALL

-- CUSTOMER_ID
SELECT
    'orders',
    'customer_id',
    COUNT(*),
    SUM(customer_id IS NULL),
    0,
    ROUND(100.0 * SUM(customer_id IS NULL) / COUNT(*), 2),
    NULL
FROM orders

UNION ALL

-- ORDER_TIME
SELECT
    'orders',
    'order_time',
    COUNT(*),
    SUM(order_time IS NULL),
    0,
    ROUND(100.0 * SUM(order_time IS NULL) / COUNT(*), 2),
    NULL
FROM orders

UNION ALL

-- PAYMENT_METHOD
SELECT
    'orders',
    'payment_method',
    COUNT(*),
    SUM(payment_method IS NULL),
    SUM(TRIM(COALESCE(payment_method, '')) = ''),
    ROUND(
        100.0 * (
            SUM(payment_method IS NULL) +
            SUM(TRIM(COALESCE(payment_method, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM orders

UNION ALL

-- DISCOUNT_PCT
SELECT
    'orders',
    'discount_pct',
    COUNT(*),
    SUM(discount_pct IS NULL),
    0,
    ROUND(100.0 * SUM(discount_pct IS NULL) / COUNT(*), 2),
    NULL
FROM orders

UNION ALL

-- TOTAL_USD
SELECT
    'orders',
    'total_usd',
    COUNT(*),
    SUM(total_usd IS NULL),
    0,
    ROUND(100.0 * SUM(total_usd IS NULL) / COUNT(*), 2),
    NULL
FROM orders

UNION ALL

-- COUNTRY
SELECT
    'orders',
    'country',
    COUNT(*),
    SUM(country IS NULL),
    SUM(TRIM(COALESCE(country, '')) = ''),
    ROUND(
        100.0 * (
            SUM(country IS NULL) +
            SUM(TRIM(COALESCE(country, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM orders

UNION ALL

-- DEVICE
SELECT
    'orders',
    'device',
    COUNT(*),
    SUM(device IS NULL),
    SUM(TRIM(COALESCE(device, '')) = ''),
    ROUND(
        100.0 * (
            SUM(device IS NULL) +
            SUM(TRIM(COALESCE(device, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM orders

UNION ALL

-- SOURCE
SELECT
    'orders',
    'source',
    COUNT(*),
    SUM(source IS NULL),
    SUM(TRIM(COALESCE(source, '')) = ''),
    ROUND(
        100.0 * (
            SUM(source IS NULL) +
            SUM(TRIM(COALESCE(source, '')) = '')
        ) / COUNT(*),
        2
    ),
    NULL
FROM orders
"""

if engine is not None:
    df_data_quality = pd.read_sql(query_data_quality, engine)

def get_status_icon(row):
    if row["missing_pct"] > 50:
        return "🔴"
    elif row["missing_pct"] > 0:
        return "⚠️"
    elif pd.notna(row["duplicate_count"]) and row["duplicate_count"] > 0:
        return "⚠️"
    else:
        return "✅"

df_data_quality["status"] = df_data_quality.apply(get_status_icon, axis=1)

display(df_data_quality)

,table_name,column_name,row_count,null_count,empty_count,missing_pct,duplicate_count,status
0,customers,customer_id,20000,0.0,0.0,0.00,0.0,✅
1,customers,signup_date,20000,0.0,0.0,0.00,NaN,✅
2,customers,country,20000,0.0,0.0,0.00,NaN,✅
3,sessions,session_id,120000,0.0,0.0,0.00,0.0,✅
4,sessions,customer_id,120000,0.0,0.0,0.00,NaN,✅
5,sessions,start_time,120000,0.0,0.0,0.00,NaN,✅
6,sessions,device,120000,0.0,0.0,0.00,NaN,✅
7,sessions,source,120000,0.0,0.0,0.00,NaN,✅
8,sessions,country,120000,0.0,0.0,0.00,NaN,✅
9,events,event_id,760958,0.0,0.0,0.00,0.0,✅


### Data Quality Assessment

The dataset is generally clean and well-structured, with minimal missing or duplicate values across key tables.

In the events table, missing values in payment and amount_usd are expected, as these attributes are only populated for purchase-related events. This is consistent with the event schema and does not indicate data quality issues.

The orders table is complete and suitable for analyzing completed transactions, including payment methods, discounts, and order values.

Overall, the dataset is sufficient for analyzing user behavior across the funnel, although it lacks detailed event-level data within the checkout process.

# Product Overview

In [5]:
query_overview = """
SELECT 
    'total_users' AS metric,
    COUNT(DISTINCT c.customer_id) AS value
FROM customers c 

UNION ALL

SELECT 
    'active_users' AS metric,
    COUNT(DISTINCT s.customer_id) AS value
FROM sessions s  

UNION ALL

SELECT 
    'total_sessions' AS metric,
    COUNT(DISTINCT s.session_id) AS value
FROM sessions s 

UNION ALL

SELECT 
    'total_orders' AS metric,
    COUNT(DISTINCT o.order_id) AS value
FROM orders o 

UNION ALL

SELECT 
    'total_buyers' AS metric,
    COUNT(DISTINCT o.customer_id) AS value
FROM orders o 
"""

overview_raw = pd.read_sql(query_overview, engine)

summary = dict(zip(overview_raw["metric"], overview_raw["value"]))

overview = pd.DataFrame({
    "metric": [
        "total_users",
        "active_users",
        "active_rate, %",
        "total_sessions",
        "sessions_per_user",
        "total_orders",
        "total_buyers",
        "conversion_rate(total users), %",
        "conversion_rate(active users), %"
    ],
    "value": [
        summary["total_users"],
        summary["active_users"],
        round(summary["active_users"] / summary["total_users"] * 100, 2),
        summary["total_sessions"],
        round(summary["total_sessions"] / summary["active_users"], 2),
        summary["total_orders"],
        summary["total_buyers"],
        round(summary["total_buyers"] / summary["total_users"] * 100, 2),
        round(summary["total_buyers"] / summary["active_users"] * 100, 2)
    ]
})

display(
    overview.style
    .hide(axis="index")
    .set_properties(**{"text-align": "left"})
    .format({"value": "{:.2f}"})
)

metric,value
total_users,20000.00
active_users,19945.00
"active_rate, %",99.72
total_sessions,120000.00
sessions_per_user,6.02
total_orders,33580.00
total_buyers,16268.00
"conversion_rate(total users), %",81.34
"conversion_rate(active users), %",81.56


## Product Overview — Summary

The product demonstrates strong user engagement and a high overall conversion rate.

Nearly all registered users are active (~99%), indicating that user activation is not a bottleneck in the funnel.

Users engage repeatedly with the product, with an average of ~6 sessions per user, suggesting sustained interest rather than one-time usage.

The overall conversion rate (~81%) is high, indicating that users who interact with the product are generally likely to complete a purchase.

## Funnel Users

In [6]:
funnel = pd.DataFrame({
    "stage": ["Total Users", "Active Users", "Buyers"],
    "users": [
        summary["total_users"],
        summary["active_users"],
        summary["total_buyers"]
    ]
})

funnel["drop_pct"] = funnel["users"].pct_change() * 100

funnel["left"] = -funnel["users"]
funnel["right"] = funnel["users"]

center_text = []
for i, row in funnel.iterrows():
    if i == 0:
        center_text.append(f"{int(row['users']):,}")
    else:
        center_text.append(f"{int(row['users']):,}<br>{row['drop_pct']:.1f}%")

fig_funnel  = go.Figure()

fig_funnel.add_trace(go.Bar(
    y=funnel["stage"],
    x=funnel["left"],
    orientation="h",
    text=None,
    hovertemplate="Stage: %{y}<br>Users: %{customdata:,}<extra></extra>",
    customdata=funnel["users"],
    name=""
))

fig_funnel.add_trace(go.Bar(
    y=funnel["stage"],
    x=funnel["right"],
    orientation="h",
    text=center_text,
    textposition="inside",
    insidetextanchor="middle",
    hovertemplate="Stage: %{y}<br>Users: %{x:,}<extra></extra>",
    name=""
))

fig_funnel.update_layout(
    title={
        'text': "User Funnel",
        'x': 0.5,
        'xanchor': 'center'
    },
    barmode="overlay",
    showlegend=False,
    xaxis=dict(
        showticklabels=False,
        title=""
    ),
    yaxis=dict(
        autorange="reversed",
        title=""
    ),
    plot_bgcolor="white",
    paper_bgcolor="white",
    bargap=0.25,
    width=900,
    height=500
)

fig_funnel.show()

### Insights
The funnel shows a very small drop from total users to active users, indicating high engagement.
However, there is a noticeable drop at the purchase stage, suggesting that not all active users convert into buyers.

In [7]:
query_events = """
SELECT
    CASE
        WHEN e.event_type = 'page_view' THEN 1
        WHEN e.event_type = 'add_to_cart' THEN 2
        WHEN e.event_type = 'checkout' THEN 3
        WHEN e.event_type = 'purchase' THEN 4
    END AS funnel_step,
    
    e.event_type,
    
    COUNT(DISTINCT s.customer_id) AS total_users,
    COUNT(DISTINCT s.session_id) AS total_sessions

FROM events e
JOIN sessions s 
    ON e.session_id = s.session_id

GROUP BY 
    funnel_step,
    e.event_type

ORDER BY funnel_step
"""

event_funnel = pd.read_sql(query_events, engine)
display(event_funnel)


,funnel_step,event_type,total_users,total_sessions
0,1,page_view,19945,120000
1,2,add_to_cart,19660,81518
2,3,checkout,17898,44909
3,4,purchase,16268,33580


In [8]:
event_funnel["drop_pct"] = event_funnel["total_users"].pct_change() * 100

event_funnel["left"] = -event_funnel["total_users"]
event_funnel["right"] = event_funnel["total_users"]

center_text = []
for i, row in event_funnel.iterrows():
    if i == 0:
        center_text.append(f"{int(row['total_users']):,}")
    else:
        center_text.append(f"{int(row['total_users']):,}<br>{row['drop_pct']:.1f}%")

fig_event_funnel  = go.Figure()

fig_event_funnel.add_trace(go.Bar(
    y=event_funnel["event_type"],
    x=event_funnel["left"],
    orientation="h",
    text=None,
    hovertemplate="Event: %{y}<br>Users: %{customdata:,}<extra></extra>",
    customdata=event_funnel["total_users"],
    name=""
))

fig_event_funnel.add_trace(go.Bar(
    y=event_funnel["event_type"],
    x=event_funnel["right"],
    orientation="h",
    text=center_text,
    textposition="inside",
    insidetextanchor="middle",
    hovertemplate="Event: %{y}<br>Users: %{x:,}<extra></extra>",
    name=""
))
fig_event_funnel.update_layout(
    title={
        'text': "User Event Funnel",
        'x': 0.5,
        'xanchor': 'center'
    },
    barmode="overlay",
    showlegend=False,
    xaxis=dict(
        showticklabels=False,
        title=""
    ),
    yaxis=dict(
        autorange="reversed",
        title=""
    ),
    plot_bgcolor="white",
    paper_bgcolor="white",
    bargap=0.25,
    width=900,
    height=500
)

fig_event_funnel.show()

In [9]:
event_funnel["drop_pct"] = event_funnel["total_sessions"].pct_change() * 100

event_funnel["left"] = -event_funnel["total_sessions"]
event_funnel["right"] = event_funnel["total_sessions"]

center_text = []
for i, row in event_funnel.iterrows():
    if i == 0:
        center_text.append(f"{int(row['total_sessions']):,}")
    else:
        center_text.append(f"{int(row['total_sessions']):,}<br>{row['drop_pct']:.1f}%")

fig_event_funnel  = go.Figure()

fig_event_funnel.add_trace(go.Bar(
    y=event_funnel["event_type"],
    x=event_funnel["left"],
    orientation="h",
    text=None,
    hovertemplate="Event: %{y}<br>Users: %{customdata:,}<extra></extra>",
    customdata=event_funnel["total_users"],
    name=""
))

fig_event_funnel.add_trace(go.Bar(
    y=event_funnel["event_type"],
    x=event_funnel["right"],
    orientation="h",
    text=center_text,
    textposition="inside",
    insidetextanchor="middle",
    hovertemplate="Event: %{y}<br>Users: %{x:,}<extra></extra>",
    name=""
))
fig_event_funnel.update_layout(
    title={
        'text': "Session Event Funnel",
        'x': 0.5,
        'xanchor': 'center'
    },
    barmode="overlay",
    showlegend=False,
    xaxis=dict(
        showticklabels=False,
        title=""
    ),
    yaxis=dict(
        autorange="reversed",
        title=""
    ),
    plot_bgcolor="white",
    paper_bgcolor="white",
    bargap=0.25,
    width=900,
    height=500
)

fig_event_funnel.show()

### Insights

While the overall user conversion appears stable, the session-level funnel highlights a significant drop-off between Add to Cart and Checkout. This suggests that users may require multiple sessions to complete a purchase.

This raises the question of whether this behavior is influenced by specific factors. To investigate this further, additional analysis will focus on segmentation by device, traffic source, and time between key funnel steps.

# Drop-off Behaiovar Analysis within Funnel

At this stage, building a full sequential funnel is not strictly necessary. Since the main drop-offs have already been identified, further analysis should focus on investigating specific transitions, particularly between add_to_cart and checkout, and between checkout and purchase.

## Session-level Funnel Analysis

### Conversion Overview

We analyzed user behavior at the session level to understand the conversion between key funnel stages.

In [10]:
funnel_query = """
SELECT
    customer_id,
    session_id,
    device,
    source,
    country,
    add_to_cart_time,
    checkout_time,
    purchase_time,
    CASE 
        WHEN purchase_time IS NOT NULL THEN 'purchase'
        WHEN checkout_time IS NOT NULL THEN 'checkout_only'
        ELSE 'no_checkout'
    END AS status,
    TIMESTAMPDIFF(MINUTE, add_to_cart_time, checkout_time) AS time_to_checkout_min
FROM (
    SELECT
        s.customer_id,
        s.session_id,
        s.device,
        s.source,
        s.country,
        MIN(CASE WHEN e.event_type = 'add_to_cart' THEN e.event_time END) AS add_to_cart_time,
        MIN(CASE WHEN e.event_type = 'checkout' THEN e.event_time END) AS checkout_time,
        MIN(CASE WHEN e.event_type = 'purchase' THEN e.event_time END) AS purchase_time
    FROM events e
    JOIN sessions s 
        ON e.session_id = s.session_id
    GROUP BY s.customer_id, s.session_id, s.device, s.source, s.country
) AS session_events
WHERE add_to_cart_time IS NOT NULL;
"""
df_funnel = pd.read_sql(funnel_query, engine)

# Add to cart → checkout
checkout_conversion = (
    df_funnel["status"].isin(["checkout_only", "purchase"]).sum() /
    len(df_funnel)
) * 100

# Checkout → purchase
purchase_conversion = (
    (df_funnel["status"] == "purchase").sum() /
    df_funnel["status"].isin(["checkout_only", "purchase"]).sum()
) * 100

print(f"The conversion rate from add_to_cart to checkout is {checkout_conversion:.2f}%.")
print(f"The conversion rate from checkout to purchase is {purchase_conversion:.2f}%.")

The conversion rate from add_to_cart to checkout is 55.09%.
The conversion rate from checkout to purchase is 74.77%.


#### Insights 
The conversion rate from add_to_cart to checkout is 55.09%, indicating a significant drop-off at this stage.

In contrast, the conversion rate from checkout to purchase is 74.77%, suggesting that users who reach checkout are more likely to complete the purchase.

This indicates that the main friction occurs before the checkout stage.

### Step-level Analysis: Add to Cart → Checkout


### Problem Identification

A noticeable drop-off is observed between the add_to_cart and checkout stages, indicating potential friction in the transition to the checkout process.

### Hypotheses

Several hypotheses may explain this behavior:

    - H1 (Device impact): Conversion from add_to_cart to checkout may differ across devices due to variations in user experience.
    - H2 (Time-related drop-off): Longer delays between add_to_cart and checkout may lead to a loss of interest or increased friction.
    - H3 (Traffic Source Impact): Conversion from add_to_cart to checkout may vary depending on the traffic source, as different channels bring users with different levels of purchase intent.
    - H4 (Country impact): Conversion from add_to_cart to checkout may vary across countries due to differences in user behavior, pricing sensitivity, or local conditions.
    - H5 (Cart size impact): Users who add more items to the cart are more likely to proceed to checkout, as this may indicate stronger purchase intent.

#### Hypothese 1
(Device impact): Conversion from add_to_cart to checkout may differ across devices due to variations in user experience.

##### Analysis Approach
Conversion rates from add_to_cart to checkout were calculated for each device type.

In [11]:
conversion_by_device = (
    df_funnel
    .groupby("device")
    .agg(
        total=("session_id", "nunique"),
        checkout=("status", lambda x: x.isin(["checkout_only", "purchase"]).sum())
    )
    .reset_index()
)
conversion_by_device["conversion_rate"] = (
    conversion_by_device["checkout"] / conversion_by_device["total"])*100

fig_by_device = px.bar(
    conversion_by_device.sort_values("total", ascending=False),
    x="device",
    y="total",
    text=conversion_by_device["conversion_rate"].round(1),
    title="Sessions by Device with Conversion Rate"
)
fig_by_device.update_layout(
    title={
        "text": "Checkout Conversion Rate by Device",
        "x": 0.5,
        "xanchor": "center"
    }
)
fig_by_device.update_layout(
    xaxis_title="Device Type",
    yaxis_title="Distribution"
)

fig_by_device.show()

#### Insights Hypothese 1:
Conversion rates from add_to_cart to checkout are consistent (~55%) across all devices, with no significant differences between platforms.

However, the volume of sessions varies notably across devices, with mobile driving the majority of user activity.

This suggests that while device type does not impact conversion efficiency, it plays an important role in overall traffic distribution.

#### Hypothese 2 
(Time-related drop-off): Longer delays between add_to_cart and checkout may lead to a loss of interest or increased friction.

##### Analysis Approach
Time-to-checkout analysis: We analyzed the distribution of time between add_to_cart and checkout using time buckets.

In [12]:
df_time = df_funnel[df_funnel["status"].isin(["checkout_only", "purchase"])].copy()

df_time["time_bucket"] = pd.cut(
    df_time["time_to_checkout_min"],
    bins=[0, 5, 15, 30, 60, 120, 1000],
    labels=["0-5", "5-15", "15-30", "30-60", "60-120", "120+"]
)

bucket_counts = df_time["time_bucket"].value_counts().sort_index()
bucket_pct = bucket_counts / bucket_counts.sum() * 100

fig_by_time = px.bar(
    x=bucket_counts.index,
    y=bucket_counts.values,
    text=bucket_pct.round(1)
)

fig_by_time.update_layout(
    title={
        "text": "Distribution of Checkout Times after Add to Cart",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Time to Checkout (minutes)",
    yaxis_title="Number of Sessions Reaching Checkout"
)

fig_by_time.show()

#### Insights Hypothese 2:
Most checkouts occur within 30–120 minutes after adding items to the cart, suggesting that users typically take some time before completing their purchase.

#### Hypothese 3 
(Traffic Source Impact): Conversion from add_to_cart to checkout may vary depending on the traffic source, as different channels bring users with different levels of purchase intent.

##### Analysis Approach
Conversion rates were calculated at the session level for each traffic source by dividing the number of sessions that reached checkout by the total number of sessions with add_to_cart.

In [13]:
conversion_by_source = (
    df_funnel
    .groupby("source")
    .agg(
        total=("session_id", "nunique"),
        checkout=("status", lambda x: x.isin(["checkout_only", "purchase"]).sum())
    )
    .reset_index()
)
conversion_by_source["conversion_rate"] = (conversion_by_source["checkout"] / conversion_by_source["total"])*100

fig_by_source = px.bar(
    conversion_by_source.sort_values("total", ascending=False),
    x="source",
    y="total",
    text=conversion_by_source["conversion_rate"].round(1),
    title="Sessions by Traffic Source with Conversion Rate"
)
fig_by_source.update_layout(
    title={
        "text": "Checkout Conversion Rate by Traffic Source",
        "x": 0.5,
        "xanchor": "center"
    }
)
fig_by_source.update_layout(
    xaxis_title="Traffic Source",
    yaxis_title="Conversion Rate (%)"
)

fig_by_source.show()

#### Insights Hypothese 3:
Conversion rates from add_to_cart to checkout are consistent (~55%) across all traffic sources, with no significant differences observed between channels.

However, the volume of sessions varies across sources. Organic and direct channels account for the largest share of user traffic, while social and referral sources contribute comparatively fewer sessions.

This suggests that although acquisition channel does not significantly impact conversion efficiency, it plays an important role in driving overall traffic and, consequently, total checkout volume.

#### Hypothese 4 
(Country impact): Conversion from add_to_cart to checkout may vary across countries due to differences in user behavior, pricing sensitivity, or local conditions.

##### Analysis Approach
To test this hypothesis, we calculated the conversion rate from add_to_cart to checkout for each country.

For each country, we measured the proportion of sessions that reached checkout among all sessions where users added items to the cart.

In [14]:
conversion_by_country = (
    df_funnel
    .groupby("country")
    .agg(
        total=("session_id", "nunique"),
        checkout=("status", lambda x: x.isin(["checkout_only", "purchase"]).sum())
    )
    .reset_index()
)

conversion_by_country["conversion_rate"] = (
    conversion_by_country["checkout"] / conversion_by_country["total"] * 100
)

fig_by_country= px.bar(
    conversion_by_country.sort_values("total", ascending=False),
    x="country",
    y="total",
    text=conversion_by_country["conversion_rate"].round(1),
    title="Sessions by Country with Conversion Rate"
)

fig_by_country.update_layout(
    title={
        "text": "Sessions by Country with Conversion Rate",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Country",
    yaxis_title="Number of Sessions"
)

fig_by_country.show()

#### Insights Hypothese 4:

Conversion rates are consistent across countries (~54–56%), indicating that geographic location does not significantly impact progression to checkout.

However, session volume varies, with the US contributing the largest share of traffic and thus driving the majority of checkout activity.

#### Hypothese 5
(Cart size impact): Users who add more items to the cart are more likely to proceed to checkout, as this may indicate stronger purchase intent.

##### Analysis Approach
To test this hypothesis, sessions were grouped based on the number of items added to the cart.

For each group, we calculated the conversion rate from add_to_cart to checkout.

In [15]:
cart_query = """
SELECT
    session_id,
    SUM(CASE WHEN event_type = 'add_to_cart' THEN qty ELSE 0 END) AS cart_items_count,
    MAX(CASE 
        WHEN event_type IN ('checkout', 'purchase') THEN 1 
        ELSE 0 
    END) AS has_checkout
FROM events
GROUP BY session_id
HAVING cart_items_count > 0;
"""

df_cart = pd.read_sql(cart_query, engine)
df_cart["cart_bucket"] = pd.cut(
    df_cart["cart_items_count"],
    bins=[0, 1, 2, 5, 10, 100],
    labels=["1", "2", "3-5", "6-10", "10+"]
)

conversion_by_cart = (
    df_cart
    .groupby("cart_bucket", observed=False)
    .agg(
        total_sessions=("session_id", "nunique"),
        checkout_sessions=("has_checkout", "sum")
    )
    .reset_index()
)

conversion_by_cart["conversion_rate"] = (
    conversion_by_cart["checkout_sessions"] / conversion_by_cart["total_sessions"] * 100
)

fig_by_cart = px.bar(
    conversion_by_cart,
    x="cart_bucket",
    y="total_sessions",
    text=conversion_by_cart["conversion_rate"].round(1).astype(str) + "%",
    title="Sessions by Cart Size with Conversion Rate"
)

fig_by_cart.update_layout(
    title={
        "text": "Sessions by Cart Size with Conversion Rate",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Number of Items in Cart",
    yaxis_title="Number of Sessions"
)

fig_by_cart.show()

#### Insights Hypothese 5:
The majority of sessions involve adding only 1 item to the cart, while larger cart sizes are significantly less common.

Conversion rates remain relatively consistent across all cart sizes (~55–56%), indicating that the number of items in the cart does not significantly impact progression to checkout.

This suggests that users tend to make quick purchase decisions regardless of cart size, and higher cart volumes do not necessarily translate into higher conversion. 

User behavior suggests a preference for single-item purchases, with limited use of the cart for bundling multiple products. This may indicate that users visit the platform with a specific purchase intent rather than browsing multiple items.

#### Key Insights

The drop-off between add_to_cart and checkout is not explained by device, traffic source, or timing factors.
The analysis of the add_to_cart → checkout funnel shows that conversion remains relatively stable (~55%) across all examined dimensions, including device type, traffic source, country, and cart size.

No significant differences were observed between segments, indicating that these factors do not meaningfully impact users’ progression to checkout.

At the same time, session volume varies across segments. Mobile devices, organic and direct traffic sources, and selected countries contribute the largest share of user activity, making them the primary drivers of overall funnel performance.

Additionally, most users add only one item to the cart, while larger cart sizes are relatively rare. The consistent conversion across cart sizes suggests that users tend to make quick, single-item purchase decisions rather than building larger baskets.

The analysis of the add_to_cart → checkout stage did not reveal any significant differences across key segments such as device, traffic source, country, or cart size.

This suggests that the observed drop-off cannot be explained by these factors. However, due to the lack of detailed event-level data within the checkout flow, it is not possible to identify the exact source of friction at this stage.

Additional data, such as intermediate checkout steps, pricing breakdown (e.g., shipping, taxes), registration requirements, and user interaction events (e.g., errors or form issues), would be required for a deeper investigation.

To further explore potential drivers of user behavior, we proceed with an order-level analysis of the checkout → purchase stage. This allows us to better understand purchase patterns and identify possible factors influencing conversion at the final stage of the funnel.

## Step-level Analysis: Checkout → Purchase

### Problem Identification

A noticeable drop-off is observed between checkout and purchase, indicating potential friction in the final stage of the conversion process.

### Hypotheses

Several hypotheses may explain this behavior:

    • H1 (Payment method behavior): The distribution of payment methods may vary across purchases, indicating user preferences and potential differences in convenience or trust between options.

    • H2 (Discount behavior): Discounts may play an important role in completed purchases, as incentives can influence user decision-making at the final stage of the funnel.

    • H3 (Order value behavior): Order values may vary across completed purchases, reflecting differences in user spending behavior.
    
    • H4 (Payment preference by order value): Payment method usage may vary depending on order value, reflecting differences in user trust or transaction preferences.
    
    • H5 (Time and purchase intent): Users who take longer to reach checkout may demonstrate lower purchase intent than those who act quickly.

#### Hypothese 1
(Payment method behavior): The distribution of payment methods may vary across purchases, indicating user preferences and potential differences in convenience or trust between options.
##### Analysis Approach
To test this hypothesis, we analyzed the distribution of payment methods across completed orders to identify user preferences at the final stage of the funnel.

In [16]:
purchase_query = """
SELECT
    order_id,
    payment_method,
    discount_pct,
    total_usd
FROM orders
WHERE total_usd IS NOT NULL
  AND total_usd > 0;
"""
df_purchase = pd.read_sql(purchase_query, engine)
purchase_by_payment = df_purchase.groupby('payment_method')['order_id'].value_counts().sum()

payment_dist = (
    df_purchase
    .groupby("payment_method")
    .agg(
        orders_count=("order_id", "nunique")
    )
    .reset_index()
)
payment_dist["share"] = (
    payment_dist["orders_count"] / payment_dist["orders_count"].sum() * 100
)

fig_payment = make_subplots(
    rows=1, cols=2,
    specs=[[{"type": "domain"}, {"type": "xy"}]],
    subplot_titles=("Payment Distribution", "Orders by Payment Method"),
    column_widths=[0.25, 0.75],
    horizontal_spacing=0.1
)

fig_payment.add_trace(
    go.Pie(
        labels=payment_dist["payment_method"],
        values=payment_dist["orders_count"],
        textinfo="label+percent"
    ),
    row=1, col=1
)

fig_payment.add_trace(
    go.Bar(
        x=payment_dist["payment_method"],
        y=payment_dist["orders_count"],
        text=payment_dist["share"].round(1).astype(str) + "%",
        textposition="auto",
        marker_color="#636EFA" 
    ),
    row=1, col=2
)

fig_payment.update_layout(
    title={
        "text": "Payment Method Analysis",
        "x": 0.5,
        "xanchor": "center"
    },
    width=1500,
    height=600
)

fig_payment.show()

#### Insights Hypothese 1:
The vast majority of orders are completed using card payments (~70%), significantly outperforming all other methods. 

Alternative payment methods such as PayPal (~15%), wallet (~10%), and cash on delivery (~5%) are used considerably less frequently.

This indicates a strong user preference for card payments, suggesting that this method is perceived as the most convenient, trusted, or accessible option at the final stage of the purchase process. 

#### Hypothese 2
(Discount behavior): Discounts may play an important role in completed purchases, as incentives can influence user decision-making at the final stage of the funnel.
##### Analysis Approach
To test this hypothesis, we analyzed the distribution of discounts across completed orders to assess how frequently incentives are used in purchase behavior.|

In [17]:
discount_dist = (
    df_purchase
    .groupby("discount_pct")
    .agg(
        orders_count=("order_id", "nunique")
    )
    .reset_index()
    .sort_values("discount_pct")
)
discount_dist["share"] = (
    discount_dist["orders_count"] / discount_dist["orders_count"].sum() * 100
)

fig_discount = px.bar(
        discount_dist,
        x="discount_pct",
        y="orders_count",
        text=discount_dist["share"].round(1).astype(str) + "%"
)

fig_discount.update_layout(
    title={
        "text": "Discount Analysis",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Discount Level (%)",
    yaxis_title="Number of Orders"
)

fig_discount.show()

#### Insights Hypothese 2:
Most purchases are completed without discounts, suggesting that users do not rely heavily on incentives to complete transactions. 

Discounts are used less frequently and show no strong dominance at any specific level.

The high share of non-discounted purchases indicates that relying heavily on discounts may not be necessary to drive conversions. Instead, users appear to be motivated by intrinsic factors such 
as product relevance or urgency.

#### Hypothese 3
(Order value behavior): Order values may vary across completed purchases, reflecting differences in user spending behavior.
##### Analysis Approach
To test this hypothesis, we analyzed the distribution of order values across completed purchases to understand typical spending patterns.

In [18]:
df_purchase["value_bucket"] = pd.cut(
    df_purchase["total_usd"],
    bins=[0, 25, 50, 100, 200, 500, 10000],
    labels=["0-25", "25-50", "50-100", "100-200", "200-500", "500+"]
)

order_value_dist = (
    df_purchase
    .groupby("value_bucket", observed=False)
    .agg(
        orders_count=("order_id", "nunique")
    )
    .reset_index()
)

order_value_dist["share"] = (
    order_value_dist["orders_count"] / order_value_dist["orders_count"].sum() * 100
)

fig_order_value = px.bar(
        order_value_dist,
        x="value_bucket",
        y="orders_count",
        text=order_value_dist["share"].round(1).astype(str) + "%"
)

fig_order_value.update_layout(
    title={
        "text": "Order Values Analysis",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Order values, $",
    yaxis_title="Number of Orders"
)

fig_order_value.show()


#### Insights Hypothese 3:
Most purchases are concentrated in the mid-range order values, particularly between $50–100 (23.6%) and $100–200 (24.8%), which together account for nearly half of all orders.

Lower-value purchases (below $50) are less frequent, while high-value orders (above $200) decline significantly, with very few purchases exceeding $500 (~3%).

This suggests that users tend to make moderate spending decisions, avoiding both very small and very large purchases at the final stage of the funnel.

The sharp decline in higher-value orders may indicate psychological or financial barriers, as users become more cautious when committing to larger purchases.

The primary friction in the funnel occurs before the checkout stage and is not driven by specific user segments. Instead, user behavior appears consistent across dimensions, suggesting a general UX or product-related issue. 

At the same time, purchase behavior indicates strong user intent, with most users completing purchases without discounts and within a moderate price range, using a dominant payment method.

#### Hypothese 4:
(Payment preference by order value): Payment method usage may vary depending on order value, reflecting differences in user trust or transaction preferences.
#### Analysis Approach
To test this hypothesis, completed orders were grouped by order value buckets, and the distribution of payment methods was compared across these groups.

In [19]:
payment_by_value = (
    df_purchase
    .groupby(["value_bucket", "payment_method"], observed=False)
    .agg(
        orders_count=("order_id", "nunique")
    )
    .reset_index()
)
payment_by_value["share_within_bucket"] = (
    payment_by_value["orders_count"] /
    payment_by_value.groupby("value_bucket", observed=False)["orders_count"].transform("sum") * 100
)
fig_payment_value = px.bar(
    payment_by_value,
    x="value_bucket",
    y="orders_count",
    color="payment_method",
    barmode="stack",
    title="Payment Method Distribution by Order Value"
)

fig_payment_value.update_layout(
    title={
        "text": "Payment Method Distribution by Order Value",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Order Value ($)",
    yaxis_title="Number of Orders"
)

fig_payment_value.show()

#### Insights Hypothese 4:
Despite differences in order value, users consistently prefer card payments, indicating a stable and dominant payment behavior across all purchase segments.

The lack of variation in payment method distribution suggests that transaction size does not significantly impact how users choose to pay, and that trust or convenience factors associated with card payments remain constant across different spending levels.

#### Hypothese 5:
(Time and purchase intent): Users who take longer to reach checkout may demonstrate lower purchase intent than those who act quickly.
#### Analysis Approach
To test this hypothesis, we analyzed sessions that reached checkout and grouped them by time to checkout. For each group, we calculated the share of sessions that resulted in a purchase.

In [20]:
purchase_by_time = (
    df_time
    .groupby("time_bucket", observed=False)
    .agg(
        total_sessions=("session_id", "nunique"),
        purchase_sessions=("status", lambda x: (x == "purchase").sum())
    )
    .reset_index()
)

purchase_by_time["purchase_rate"] = (
    purchase_by_time["purchase_sessions"] / purchase_by_time["total_sessions"] * 100
)
fig_time_purchase = px.bar(
    purchase_by_time,
    x="time_bucket",
    y="purchase_rate",
    text=purchase_by_time["purchase_rate"].round(1).astype(str) + "%",
    title="Purchase Rate by Time to Checkout"
)

fig_time_purchase.update_traces(
    textposition="auto",
    marker_color="#636EFA"
)

fig_time_purchase.update_layout(
    title={
        "text": "Purchase Rate by Time to Checkout",
        "x": 0.5,
        "xanchor": "center"
    },
    xaxis_title="Time to Checkout (minutes)",
    yaxis_title="Purchase Rate (%)"
)

fig_time_purchase.show()

#### Insights Hypothese 5:
Despite differences in the time users take to reach checkout, purchase rates remain stable across all time groups. This suggests that checkout timing is not a strong indicator of purchase intent, and users who reach checkout are equally likely to complete the purchase regardless of how quickly or slowly they progress through the funnel.

## Key Insights:
The analysis suggests that neither user segmentation (device, traffic source, country) nor behavioral timing significantly explain the drop-off. This indicates that the main friction likely occurs within the checkout experience itself and requires deeper investigation.

# Final Conclusion
This analysis examined user behavior across the e-commerce funnel, focusing on the transition from add_to_cart to checkout and from checkout to purchase.

The results show that the main drop-off occurs before the checkout stage, with only ~~55% of users proceeding from add_to_cart to checkout. In contrast, the checkout to purchase conversion is relatively high (~74%), indicating that users who reach checkout are generally likely to complete their purchase.

Multiple factors were tested at the session level, including device type, traffic source, country, cart size, and time to checkout. None of these variables showed a significant impact on conversion, suggesting that user segmentation and behavioral timing do not explain the observed drop-off.

Order-level analysis (payment method, discount usage, and order value) provided insights into user preferences but could not explain conversion behavior, as this data is only available for completed purchases.

Overall, the findings indicate that the primary friction likely occurs at or before the entry point to checkout. This may be driven with factors such as unclear pricing (e.g., shipping or taxes), lack of trust, or hesitation before committing to the purchase.

To address this issue, further investigation is required with more granular data, including detailed checkout flow events, pricing breakdowns, and user interaction signals (e.g., errors or form issues).

Potential improvement areas include:
- simplifying checkout steps  
- reducing mandatory fields  
- displaying full pricing earlier  
- optimizing the checkout experience across devices  

In conclusion, while user characteristics do not explain the conversion gap, the analysis highlights a clear need to focus on product and checkout experience optimization.